### ISSUE TITLE
3d Graphview python file intergrate with andriod studio i am facing some issue

### ISSUE THREAD

**Issue by deepik27** [chaquo/chaquopy#474]
Hi, 

i want to integrate python 3d surface plot to android studio, i used graph view but i am facing some issues also can give me some few ideas i will attach python code here

from mpl_toolkits.mplot3d import Axes3D
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

# Dataset
df=pd.DataFrame({'X': range(1,101), 'Y': np.random.randn(100)*15+range(1,101), 'Z': (np.random.randn(100)*15+range(1,101))*2 })

from mpl_toolkits.mplot3d import Axes3D
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

# Dataset
df=pd.DataFrame({'X': range(1,101), 'Y': np.random.randn(100)*15+range(1,101), 'Z': (np.random.randn(100)*15+range(1,101))*2 })

# plot
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.scatter(df['X'], df['Y'], df['Z'], c='skyblue', s=60)
ax.view_init(30, 185)
plt.show()

**Comment by mhsmith** [comment]
And what's your issue?

**Comment by mhsmith** [comment]
You should return the image to Java/Kotlin code and then display it using an ImageView. See [chaquopy-matplotlib](https://github.com/chaquo/chaquopy-matplotlib) for an example.

**Comment by deepik27** [comment]
Actually it is 3d surface plot written in python with csv file.....only one
frame I can return to in andriod studio....should I use graphview ?.. how
to use graphview?

**Comment by mhsmith** [comment]
> only one frame I can return to in andriod studio

I don't know what you mean by that. If you want to display multiple frames, then put your code in a function and call it multiple times, like the chaquopy-matplotlib example I linked above.

> should I use graphview ?.. how to use graphview?

I've never used [GraphView](https://github.com/jjoe64/GraphView), but since it's written specifically for Android I expect it would give better performance than Matplotlib, especially for interactive use. You can either call it from Java/Kotlin code as described in GraphView's own documentation, or from Python using the [Chaquopy Python API](https://chaquo.com/chaquopy/doc/current/python.html).

**Comment by deepik27** [comment]
I am getting this error, 
W/python.stderr: Unlicensed copy of Chaquopy: app will now shut down. See https://chaquo.com/chaquopy/license/
W/python.stderr:  



A/libc: Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0 in tid 7872 (Thread-2), pid 7786 (ntegrate_python)

**Comment by deepik27** [comment]
Actually i want to implement like this 3d graph in andriod studio, In python i wrote all code its working also in that, how can i integrate i used ur matplolib code, but its not working can u give me some best solution for this thanks in advance
![Screenshot from 2021-04-05 12-27-30](https://user-images.githubusercontent.com/64435256/113547050-d0b08200-960a-11eb-9511-e80321178ab4.png)
![Screenshot from 2021-04-05 12-27-13](https://user-images.githubusercontent.com/64435256/113547051-d1e1af00-960a-11eb-9676-0f836e0bf5c6.png)
![Screenshot from 2021-04-05 12-27-08-1](https://user-images.githubusercontent.com/64435256/113547054-d27a4580-960a-11eb-99f8-c2d9bd42a5b7.png)
![Screenshot from 2021-04-05 12-27-08](https://user-images.githubusercontent.com/64435256/113547055-d312dc00-960a-11eb-95ce-1b9df9017cc0.png)
![Screenshot from 2021-04-05 12-27-04](https://user-images.githubusercontent.com/64435256/113547056-d3ab7280-960a-11eb-9805-f49d068fb65c.png)

**Comment by deepik27** [comment]
[1-0-0.txt](https://github.com/chaquo/chaquopy/files/6257308/1-0-0.txt)

This is my csv file and code is here below




import numpy as np
import pandas as pd
import scipy.io
import cv2
from PIL import Image
import matplotlib
import matplotlib.pyplot as plt
from matplotlib import cm
from mpl_toolkits.mplot3d import Axes3D
from matplotlib.colors import LightSource
from csv import reader
from matplotlib import style
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from os.path import dirname, join
import io
import base64

def main ():

    csvPath= join(dirname(__file__),"1-0-0.csv")
    df = pd.read_csv(csvPath,sep=",",skiprows=1)
    df = df.drop(df.columns[0], axis=1)
    df = df.iloc[:,:-6]
    print(df.info())
    xor = df.values #returns a numpy array
    min_max_scaler = preprocessing.MinMaxScaler()
    x_scaled = min_max_scaler.fit_transform(xor)
    df = pd.DataFrame(x_scaled)
    print(df.shape)
    row = 23
    column = 31
    frame=[]
    fig = plt.figure(figsize=(15,8))
    plt.ion()
    ax1 = fig.add_subplot(111, projection='3d')
    i=0
    while True:
        try:
            df_np=df.iloc[i].values
            frame = df_np.reshape(row, column)
            matrix=frame
            (X, Y) = np.meshgrid(np.arange(matrix.shape[0]), np.arange(matrix.shape[1]))
            print(X,Y)
            ax1.set_title('SureTouch Solid Model-Surface Plots')
            ax1.set_xlabel('x-axis')
            ax1.set_ylabel('y-axis')
            ax1.set_zlabel('z-axis')
            surf=ax1.plot_surface(X, Y, np.transpose(matrix),cmap=plt.cm.coolwarm)
            plt.show()
            plt.pause(0.1)
            ax1.cla()
            fig.canvas.draw()
            img = np.fromstring(fig.canvas.tostring_rgb(),dtype=np.uint8,sep='')
            img = img.reshape(fig.canvas.get_width_height()[::-1]+(3,))
            img=cv2.cvtColor(img,cv2.COLOR_RGB2BGR)
            pil_im = Image.fromarray(img)
            buff = io.BytesIO()
            pil_im.save(buff, format="PNG")
            img_str = base64.b64encode(buff.getvalue())
            print("[INFO]plotting frame=",i)
            i+=1
        except IndexError:
            print("\n[INFO]End---No data available")
            break
    return ""+str(img_str, 'utf-8')


Can you please give some idea how to implement this it will be great helpful for me now, i am using java in android app

**Comment by deepik27** [comment]
This is my mainactivity.java file:



package com.example.integrate_python;

import androidx.appcompat.app.AppCompatActivity;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.util.Base64;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.chaquo.python.PyObject;
import com.chaquo.python.Python;
import com.chaquo.python.android.AndroidPlatform;
import com.google.android.material.bottomnavigation.BottomNavigationItemView;
import com.jjoe64.graphview.GraphView;
import com.jjoe64.graphview.series.DataPoint;
import com.jjoe64.graphview.series.LineGraphSeries;
import android.widget.Toast;
public class MainActivity extends AppCompatActivity {


    Button btn;
    ImageView iv;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        btn = (Button)findViewById(R.id.submit);
        iv = (ImageView)findViewById(R.id.image);


        //Now checking python is started or not
        if(!Python.isStarted())
            Python.start(new AndroidPlatform(this));

        final Python py = Python.getInstance();

        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {


                //python object to load the script

                PyObject pyo = py.getModule("meshgrid_frameall");
                PyObject obj = pyo.callAttr("main");
                String str = obj.toString();
                byte data[] = android.util.Base64.decode(str, Base64.DEFAULT);
                Bitmap bmp = BitmapFactory.decodeByteArray(data, 0, data.length);
                iv.setImageBitmap(bmp);

            }

        });

            }



}

**Comment by deepik27** [comment]
Actually i am going to implement live graph so ? please help me out

**Comment by mhsmith** [comment]
> Unlicensed copy of Chaquopy: app will now shut down. See https://chaquo.com/chaquopy/license/

If you want to remove the 5-limit time limit, contact us by following the instructions on that page.

> only one frame I can return to in andriod studio

That's because your code is only returning the last frame in the loop. If you want to display multiple frames then you'll need to either:

* Return all the frames in a list, then access that list from Java using [`asList`](https://chaquo.com/chaquopy/doc/current/java/com/chaquo/python/PyObject.html#asList()). If you want an animation effect, you'll need to display the frames at regular intervals and return control to the app's main loop each time to give it a chance to display. The simplest way to do this is to use [`postDelayed`](https://developer.android.com/reference/android/os/Handler#postDelayed(java.lang.Runnable,%20long)), which you can lots of examples of elsewhere.

OR

* Restructure your Python function so that instead of generating all frames in a single call, it gets called multiple times and returns a different frame each time, such as by taking `i` as a function argument. You can use `postDelayed` to make the calls at regular intervals, then display each frame as soon as it's returned.

**Comment by deepik27** [comment]
Can you show me one example so it will be useful

**Comment by mhsmith** [comment]
Sorry, I can't write your app for you. I've given you enough links to learn what you need.

**Comment by deepik27** [comment]
No i didnt asked to write a code for my app, just how to intergrate with python postdelayed for that one example i asked thats all

**Comment by deepik27** [comment]
> > Unlicensed copy of Chaquopy: app will now shut down. See https://chaquo.com/chaquopy/license/
> 
> If you want to remove the 5-limit time limit, contact us by following the instructions on that page.
> 
> > only one frame I can return to in andriod studio
> 
> That's because your code is only returning the last frame in the loop. If you want to display multiple frames then you'll need to either:
> 
> * Return all the frames in a list, then access that list from Java using [`asList`](https://chaquo.com/chaquopy/doc/current/java/com/chaquo/python/PyObject.html#asList()). If you want an animation effect, you'll need to display the frames at regular intervals and return control to the app's main loop each time to give it a chance to display. The simplest way to do this is to use [`postDelayed`](https://developer.android.com/reference/android/os/Handler#postDelayed(java.lang.Runnable,%20long)), which you can lots of examples of elsewhere.
> 
> OR
> 
> * Restructure your Python function so that instead of generating all frames in a single call, it gets called multiple times and returns a different frame each time, such as by taking `i` as a function argument. You can use `postDelayed` to make the calls at regular intervals, then display each frame as soon as it's returned.

Thank you so much for your help...it worked me

**Comment by mhsmith** [comment]
Great, please create separate issues for any future problems.
